test/bio_printf_test.c

```diff
@@ -0,0 +1,118 @@
+#include <assert.h>
+#include <stdio.h>
+
+#include <openssl/bio.h>
+#include <test/testutil.h>
+
+#if defined OPENSSL_SYS_WINDOWS
+#define TEST_HAVE_FMEMOPEN 0
+#else
+#define TEST_HAVE_FMEMOPEN 1
+#endif
+
+#define lengthof(a) (sizeof (a) / sizeof *(a))
+#define endof(a) (&(a)[lengthof(a)])
+
+#define TEST_MAXBUF 3
+
+static struct test {
+    size_t bufsiz;
+    const char *fmt;
+    int arg;
+    int expect;
+} tests[] = {
+    { 3, "%d", 1, 1 },
+    { 3, "%d", 11, 2 },
+    { 3, "%d", 111, -1 },
+#if !defined OPENSSL_SYS_WINDOWS
+    /*
+     * Windows doesn't return -1 on bad wchar_t, continuing to process the
+     * remaining format specification (though it does set EILSEQ). In this
+     * case it returns 0. Just skip rather than accommodate in test logic.
+     */
+    { 3, "%lc", -2, -1 },
+#endif
+};
+
+#if TEST_HAVE_FMEMOPEN
+static int
+test_bio_vprintf(void)
+{
+    char buf[TEST_MAXBUF];
+    int pass = 0, fail = 0;
+    for (struct test *_T = tests; _T < endof(tests); _T++) {
+        /* adjust test to account for lack of NUL involvement */
+        struct test T = *_T;
+        T.bufsiz--;
+
+        assert(sizeof buf >= T.bufsiz);
+        FILE *fp = fmemopen(buf, T.bufsiz, "w");
```

Comment:
This should have error checking so that the test throws an error rather than segfault later in the unlikely event of allocation failure/other errors.

Likewise for a couple of other calls below.